Optimize the parsing of special floats: infinity and nan.

ludi317 opened the pull request:
Don't repeat the check on the first character in each string: 'i' or 'n'. Add a benchmark for special floats.
Supplement units test on special floats with SyntaxError cases.

Benchmarks on my M1 Mac show a -8% speedup:

$ benchstat old.txt new.txt
goos: darwin
goarch: arm64
pkg: github.com/sugawarayuuta/refloat
                               │   old.txt   │              new.txt              │
                               │   sec/op    │   sec/op     vs base              │
ParseFloat64Special/refloat-10   8.895n ± 1%   8.197n ± 1%  -7.84% (p=0.002 n=6)

ludi317 committed:
Optimize the parsing of special floats: infinity and nan.
Don't repeat the check on the first character in each string: 'i' or 'n'.
Add a benchmark for special floats.
Supplement units test on special floats with SyntaxError cases.

Benchmarks on my M1 Mac show a -8% speedup:

$ benchstat old.txt new.txt
goos: darwin
goarch: arm64
pkg: github.com/sugawarayuuta/refloat
                               │   old.txt   │              new.txt              │
                               │   sec/op    │   sec/op     vs base              │
ParseFloat64Special/refloat-10   8.895n ± 1%   8.197n ± 1%  -7.84% (p=0.002 n=6)

sugawarayuuta commented:
Hello. Thank you for your contribution. I think that it might be better to modify the `float32.go` file at the same time to keep it consistent. (And adding 32bit versions to the benchmarks like you did with the other one certainly helps too.) 
I did a benchmark myself and the improvement is there although it's not that significant (~4%?). Overall I think it's a good change.

ludi317 committed:
Optimize the special floats (infinity, nan) for float32s too.
Add benchmarks for them, showing a more modest -4% improvement.

$ benchstat old.txt new.txt
goos: darwin
goarch: arm64
pkg: github.com/sugawarayuuta/refloat
                               │   old.txt   │              new.txt              │
                               │   sec/op    │   sec/op     vs base              │
ParseFloat32Special/refloat-10   7.070n ± 0%   6.812n ± 1%  -3.64% (p=0.002 n=6)

ludi317 commented:
@sugawarayuuta Done.

sugawarayuuta commented:
Thanks again!
